File: tools/links/render_check.js
// Render check: opens the artifact page from the filesystem at three viewport widths
// and reports horizontal overflow and console errors. Adopted from a sibling practice's
// transferable finding: read the page as a picture, not only as code.
//
// Usage: NODE_PATH=/opt/node22/lib/node_modules node tools/links/render_check.js <file.html>
const { chromium } = require('playwright');

(async () => {
  const file = 'file://' + require('path').resolve(process.argv[2]);
  const browser = await chromium.launch();
  let bad = 0;
  for (const width of [390, 768, 1280]) {
    const page = await browser.newPage({ viewport: { width, height: 900 } });
    const errors = [];
    page.on('console', m => { if (m.type() === 'error') errors.push(m.text()); });
    page.on('pageerror', e => errors.push(String(e)));
    const requests = [];
    page.on('request', r => { if (!r.url().startsWith('file://')) requests.push(r.url()); });
    await page.goto(file, { waitUntil: 'load' });
    const m = await page.evaluate(() => ({
      scroll: document.documentElement.scrollWidth,
      client: document.documentElement.clientWidth,
      wide: [...document.querySelectorAll('main *')]
        .filter(el => el.getBoundingClientRect().right > document.documentElement.clientWidth + 1)
        .map(el => el.tagName + '.' + (el.className || '')).slice(0, 6),
      text: document.body.innerText.length,
    }));
    const overflow = m.scroll > m.client + 1;
    if (overflow || errors.length || requests.length) bad++;
    console.log(`${width}px  overflow=${overflow} (${m.scroll}>${m.client})  ` +
      `console_errors=${errors.length}  network_requests=${requests.length}  text_chars=${m.text}`);
    if (m.wide.length) console.log('   wide elements:', m.wide.join(', '));
    errors.slice(0, 3).forEach(e => console.log('   error:', e.slice(0, 160)));
    requests.slice(0, 3).forEach(r => console.log('   network:', r.slice(0, 160)));
    await page.screenshot({ path: `/tmp/render-${width}.png`, fullPage: false });
    await page.close();
  }
  await browser.close();
  process.exit(bad ? 1 : 0);
})();

File: tools/written-by-the-agent/render_check.py
import json, os, sys
from playwright.sync_api import sync_playwright
ART = sys.argv[1]
url = "file://" + os.path.join(ART, "index.html")
out = {"note": "Rendered in a real browser from the filesystem, scripting on and off. "
               "The page carries no JavaScript, so the two must be identical.", "widths": []}
with sync_playwright() as p:
    for js in (True, False):
        b = p.chromium.launch(executable_path="/opt/pw-browsers/chromium-1194/chrome-linux/chrome")
        for w in (390, 768, 1280):
            ctx = b.new_context(viewport={"width": w, "height": 900}, java_script_enabled=js)
            pg = ctx.new_page()
            errs, reqs = [], []
            pg.on("console", lambda m: errs.append(m.text) if m.type == "error" else None)
            pg.on("pageerror", lambda e: errs.append(str(e)))
            pg.on("request", lambda r: reqs.append(r.url) if not r.url.startswith("file:") else None)
            pg.goto(url, wait_until="load")
            m = pg.evaluate("({sw: document.documentElement.scrollWidth,"
                            " iw: window.innerWidth,"
                            " tables: document.querySelectorAll('table').length,"
                            " rows: document.querySelectorAll('tbody tr').length,"
                            " controls: document.querySelectorAll("
                            "'input,select,button,textarea,details,[onclick]').length,"
                            " text: document.body.innerText.length})")
            out["widths"].append({"width": w, "javascript": js, "scroll_width": m["sw"],
                                  "inner_width": m["iw"],
                                  "horizontal_overflow": m["sw"] > m["iw"],
                                  "tables": m["tables"], "table_rows": m["rows"],
                                  "interactive_controls": m["controls"],
                                  "text_chars": m["text"],
                                  "console_errors": len(errs), "network_requests": len(reqs)})
            ctx.close()
        b.close()
json.dump(out, open(os.path.join(ART, "data", "render-check.json"), "w", encoding="utf-8"),
          indent=1, ensure_ascii=False)
for r in out["widths"]:
    print(f"  js={str(r['javascript']):5} w={r['width']:>5}  overflow={r['horizontal_overflow']}"
          f"  rows={r['table_rows']}  controls={r['interactive_controls']}"
          f"  errors={r['console_errors']}  requests={r['network_requests']}"
          f"  text={r['text_chars']}")
